UDPEndPointImplLwIP: Support LWIP_TCPIP_CORE_LOCKING=0

rojer opened the pull request:
Wrap calls to LwIP APIs in `tcpip_api_call()`, as required. When `LWIP_TCPIP_CORE_LOCKING` is enabled, this internally becomes `LOCK_TCPIP_CORE/UNLOCK_TCPIP_CORE` and when it isn't, it posts a message to the TCPIP task to run the function.

https://github.com/project-chip/connectedhomeip/issues/28590

cc @shubhamdp

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=29057) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=29057) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=29057) it.</sub>

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -35,16 +37,14 @@
 #include <lwip/ip.h>
 #include <lwip/mld6.h>
 #include <lwip/netif.h>
+#include <lwip/priv/tcpip_priv.h>
```
That does not look like public API...  the docs for that header say:

> TCPIP API internal implementations (do not use in application code)

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -35,16 +37,14 @@
 #include <lwip/udp.h>
 
-static_assert(LWIP_VERSION_MAJOR > 1, "CHIP requires LwIP 2.0 or later");
+// #include <platform/PlatformManager.h>
```
Why is this commented-out thing here?

rojer replied:
please see below

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -67,27 +67,34 @@ static_assert(LWIP_TCPIP_CORE_LOCKING, "CHIP requires config LWIP_TCPIP_CORE_LOC
+
+static void LockChipStack()
+{
+    // DeviceLayer::PlatformMgr().LockChipStack();
```
This really does not make sense.

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -113,10 +120,7 @@ CHIP_ERROR UDPEndPointImplLwIP::BindImpl(IPAddressType addressType, const IPAddr
 
 CHIP_ERROR UDPEndPointImplLwIP::BindInterfaceImpl(IPAddressType addrType, InterfaceId intfId)
 {
-    // A lock is required because the LwIP thread may be referring to intf_filter,
-    // while this code running in the Inet application is potentially modifying it.
```
Are the comments here wrong?

rojer replied:
yes, they don't apply anymore.

bzbarsky-apple replied:
Is that because `LwIPBindInterface` now does `udp_bind_netif` in a synchronized way?

Are we guaranteed that mUDP cannot change between `GetPCB` and the use of the pointer in `LwIPBindInterface`?

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -100,7 +107,7 @@ CHIP_ERROR UDPEndPointImplLwIP::BindImpl(IPAddressType addressType, const IPAddr
     if (res == CHIP_NO_ERROR)
     {
-        res = chip::System::MapErrorLwIP(udp_bind(mUDP, &ipAddr, port));
+        res = chip::System::MapErrorLwIP(RunOnTCPIP([this, &ipAddr, port]() { return udp_bind(mUDP, &ipAddr, port); }));
```
The scope of what gets run under the lock was much larger before.  Why is the scope narrowing ok?  This applies in a bunch of places.

rojer replied:
this is the right way. acquiring the lock or hopping on the TCPIP thread is only needed for the duration of the LwIP API call.

bzbarsky-apple replied:
I think that this could use a more thorough analysis than "this is the right way"....  Especially given the specific problem in GetPCB I pointed out below.  Someone needs to carefully check that none of these other functions have similar problems and put the reasons why they don't in the commit message.

rojer replied:
sure, i responded to this comment. in short, "core locking" is for the benefit of lwip, not our endpoint. if we need to protect endpoint's internal state, we should have a separate lock.

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -140,14 +144,15 @@ CHIP_ERROR UDPEndPointImplLwIP::LwIPBindInterface(struct udp_pcb * aUDP, Interfa
 
 InterfaceId UDPEndPointImplLwIP::GetBoundInterface() const
 {
-    ScopedLwIPLock lwipLock;
```
Why is synchronization here no longer needed?

rojer replied:
RunOnTCPIP is the synchronization.

bzbarsky-apple replied:
But this function is not calling RunOnTCPIP anywhere.

rojer replied:
yes, my bad, should have wrapped `netif_get_by_index`. done.

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -307,20 +311,24 @@ void UDPEndPointImplLwIP::HandleDataReceived(System::PacketBufferHandle && msg,
 
 CHIP_ERROR UDPEndPointImplLwIP::GetPCB(IPAddressType addrType)
 {
-    // IMPORTANT: This method MUST be called with the LwIP stack LOCKED!
```
Was this controlling access to mUDP via the lock?  Put another way, can two racing calls to this now both enter the "mUDP == null" block in parallel?  Seems like they can.

rojer replied:
locking the lwip stack is done to protect lwip's internal state, i.e. in this case the target of `mUDP`. it incidentally also protects mUDP itself, but that was not the primary purpose of https://github.com/project-chip/connectedhomeip/pull/28798 where it was added. i am not familiar enough with CHIP to know if it should. can endpoint construction/destruction be racy? can we have a message being sent while an endpoing is being destroyed? i was under the impression that it CHIP serializes everything to its event loop task, so it's not racy in that way. is it?

bzbarsky-apple replied:
Matter itself should not be racy that way, as long as all accesses are on the Matter thread.

If that's meant to be a guarantee here, this function should `assertChipStackLockedByCurrentThread()`.

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -370,14 +378,19 @@ void UDPEndPointImplLwIP::LwIPReceiveUDPMessage(void * arg, struct udp_pcb * pcb
 {
     UDPEndPointImplLwIP * ep = static_cast<UDPEndPointImplLwIP *>(arg);
+
+    LockChipStack();
```
Why are these new locks here?  Especially when they are no-ops?

rojer replied:
@bzbarsky-apple please see https://github.com/project-chip/connectedhomeip/issues/28590#issuecomment-1705582187 (sorry, i should have commented here). TLDR is: i think we should, but at the moment we can't. i'll need guidance here on how to proceed.

bzbarsky-apple replied:
@rojer This method was designed to be able to run from outside the Matter thread, without locking, because that's how LwIP calls it.  If it's not doing that correctly, it should be fixed, instead of adding locks...

rojer replied:
yes, LwIPReceiveUDPMessage will be invoked on another thread and since it pokes at the endpoint's internals, the endpoint's reference counter needs to be protected. i see that packet filter is also being invoked, so it should be protected too.
but how to do that other than by adding a lock (or locking the stack, as i attempted to do)? i'm confused.

bzbarsky-apple replied:
Packet filter is expected to be called from random threads; it documents that explicitly in its API documentation.  So the synchronization to be done there, if any, is the packet filter's responsibility.

The refcount of the endpoint is not touched in LwIPReceiveUDPMessage.

`mDelayReleaseCount` is an `std::atomic` precisely to avoid worrying about races on it.  That said, we do need to make sure we can't be inside LwIPReceiveUDPMessage on a different thread at the time when the `mDelayReleaseCount` check in `UDPEndPointImplLwIP::CloseImpl` happens.  I don't know whether the earlier completion of the `udp_remove` call there guarantees that, but I would hope so...  If it does not, we need to do something to synchronize here.

rojer replied:
ok, thanks for the explanation! i think we're good: `UDPEndPointImplLwIP::CloseImpl` runs `udp_remove` on the TCPIP task (or with core lock), after that no more receive callbacks will be delivered for that PCB. hence, we don't need additional synchrnization after all, so i removed CHIP stack locking calls.

wqx6 commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -67,27 +67,34 @@ static_assert(LWIP_TCPIP_CORE_LOCKING, "CHIP requires config LWIP_TCPIP_CORE_LOC
+err_t RunOnTCPIP(std::function<err_t()> fn)
+{
+    TCPIPCallContext ctx = { .fn = fn };
+    return tcpip_api_call([](struct tcpip_api_call_data * call) { return reinterpret_cast<TCPIPCallContext *>(call)->fn(); },
```
May we use the public API `tcpip_callback` instead of `tcpip_api_call` here?

rojer replied:
not quite: `tcpip_callback` just schedules a function to run on the TCPIP task, `tcpip_api_call` also waits for it to run and returns the result.
it's a fair point that we should not be using internal APIs, so i implemented `RunOnTCPIP` in the spirit of `tcpip_api_call` but without actually using it.

rojer commented:
@bzbarsky-apple ok, it's all green now, PTAL.
i realized that there is a TCP endpoint and it needs similar treatment. i'm not going to do it now (we don't use it in our project) but i moved the assertion there from the UDP endpoint and added a TODO.

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -66,27 +63,63 @@ static_assert(LWIP_TCPIP_CORE_LOCKING, "CHIP requires config LWIP_TCPIP_CORE_LOC
-class ScopedLwIPLock
+EndpointQueueFilter * UDPEndPointImplLwIP::sQueueFilter = nullptr;
+
+err_t EndPointStateLwIP::RunOnTCPIPRet(std::function<err_t()> fn)
```
This should really live in EndPointStateLwIP.cpp...

rojer replied:
ideally yes but given how build rules are paramterized, i'd also need to add empty files for other `${chip_system_config_inet}` values: `EndPointStateSockets.cpp`, `EndPointStateNetworkFramework.cpp`, `EndPointStateOpenThread.cpp` - all empty.
my reasoning was that UDP endpoint is mandatory and TCP is optional, so we can get away with placing this small common bit here, rather than have multiple empty files like that.
but if you'd like me to do that, let me know.

bzbarsky-apple replied:
I'd think you could just add the new file inside `if (chip_system_config_use_lwip) {` in the BUILD.gn.  That's the condition under which `chip_system_config_inet = "LwIP"` happens....

rojer replied:
doh! again, please excuse my unfamiliarity with the build system, it's new to me. done.

bzbarsky-apple replied:
No need to apologize! Build systems are always a pain.

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -66,27 +63,63 @@ static_assert(LWIP_TCPIP_CORE_LOCKING, "CHIP requires config LWIP_TCPIP_CORE_LOC
+    }
 
-EndpointQueueFilter * UDPEndPointImplLwIP::sQueueFilter = nullptr;
+    static std::function<err_t()> sTCPIPFunction;
+    static err_t sTCPIPFunctionResult;
```
I assume the goal here with these statics is to not have the lambda closing over stack variables to avoid heap-allocations?  If so, please document that.  If that's _not_ the goal, it's even more important to document what the goal is.

rojer replied:
the reason is that the lambda passed to `tcpip_callback()` is not really a lambda but a C function pointer. lambda syntax in this case is just syntactic sugar, it's not a real lambda and we can't really capture anything.
we could allocate a temporary struct to hold these and pass as context, i thought two small statics would be ok for our purpose, especially since we `assertChipStackLockedByCurrentThread` already and don't expect this to be invoked concurrently. i added a comment.

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -240,30 +265,29 @@ CHIP_ERROR UDPEndPointImplLwIP::SendMsgImpl(const IPPacketInfo * pktInfo, System
 {
-    ScopedLwIPLock lwipLock;
-
     // Since UDP PCB is released synchronously here, but UDP endpoint itself might have to wait
```
This should probably `assertChipStackLockedByCurrentThread`.

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -306,20 +330,18 @@ void UDPEndPointImplLwIP::HandleDataReceived(System::PacketBufferHandle && msg,
 {
-    // IMPORTANT: This method MUST be called with the LwIP stack LOCKED!
-
     // If a PCB hasn't been allocated yet...
```
This still needs assertChipStackLockedByCurrentThread(), no?

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -369,6 +391,7 @@ void UDPEndPointImplLwIP::LwIPReceiveUDPMessage(void * arg, struct udp_pcb * pcb
                                                 u16_t port)
 {
     UDPEndPointImplLwIP * ep = static_cast<UDPEndPointImplLwIP *>(arg);
+
```
```suggestion
```

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
Does FindNetifFromInterfaceId not require synchronization?  Why not?

rojer reviewed (commented):
PTAL

bzbarsky-apple commented on src/inet/UDPEndPointImplLwIP.cpp:
```diff
@@ -66,27 +63,64 @@ static_assert(LWIP_TCPIP_CORE_LOCKING, "CHIP requires config LWIP_TCPIP_CORE_LOC
+    }
 
-EndpointQueueFilter * UDPEndPointImplLwIP::sQueueFilter = nullptr;
+    // These are static to be able to share them with the TCPIP callback.
```
```suggestion
    // tcpip_callback takes a C function pointer, so we can't pass a capturing lambda to it.
    // Just store the state the function we pass to it needs in statics.
    // This should be safe, since that function will execute before we return and there is no
    // re-entry into this method.
```

bzbarsky-apple reviewed (approved):
Approved, but adding the new .cpp file (here or in a followup) would be good.
